### ISSUE TITLE
Sever IREE repo connection with mlir-hlo

### ISSUE THREAD

**Issue by julianwa** [iree-org/iree#12678]
Issues with IREE's mlir-hlo dependency have been blocking integrates for 2 weeks. We've shared the intention to cut ties with this dependency, and now we need to do so expediently to unblock the integrates.

Very high-level description of what's needed:
- Create proposal, making sure to confirm with @GMNGeoffrey and @jpienaar that the proposal is coherent when integrating IREE back into Google.
- Extract some files out of mlir-hlo, organize them appropriately in IREE, delete submodule. 
- Submit when tests pass. 

More detailed plan should be discussed with Magnus Hyttsten (not sure his github handle).

**Comment by aaron-schneider** [comment]
Hi - double checking on this P0 issue. More to say? Ok to close or lower priority? Thanks!

**Comment by GMNGeoffrey** [comment]
@kuhar is actively working on this, but I would say it's not P0 and requires some consideration.

**Comment by kuhar** [comment]
Status update:

I ported the following the lowering patterns (mhlo -> linalg) for the following stablehlo ops:
-  AbsOp
-  AddOp
-  AndOp
-  Atan2Op
-  BitcastConvertOp
-  BroadcastInDimOp
-  BroadcastOp
-  CbrtOp
-  CeilOp
-  ClampOp
-  ClzOp
-  CompareOp
-  ComplexOp
-  ConcatenateOp
-  ConstantOp
-  ConvertOp
-  ConvolutionOp
-  CosineOp
-  DivOp
-  DotGeneralOp
-  DotOp
-  DynamicSliceOp
-  DynamicUpdateSliceOp
-  EinsumOp
-  Expm1Op
-  ExpOp
-  FloorOp
-  ImagOp
-  IotaOp
-  IsFiniteOp
-  Log1pOp
-  LogisticOp
-  LogOp
-  MaxOp
-  MinOp
-  MulOp
-  NegOp
-  NotOp
-  OrOp
-  PadOp
-  PopulationCountOp
-  PowOp
-  RealDynamicSliceOp
-  RealOp
-  ReduceOp
-  ReducePrecisionOp
-  ReduceWindowOp
-  RemOp
-  ReshapeOp
-  ReturnOp
-  ReverseOp
-  RngBitGeneratorOp
-  RngOp
-  RoundNearestEvenOp
-  RoundOp
-  RsqrtOp
-  SelectAndScatterOp
-  SelectOp
-  SetDimensionSizeOp
-  ShiftLeftOp
-  ShiftRightArithmeticOp
-  ShiftRightLogicalOp
-  SignOp
-  SineOp
-  SliceOp
-  SqrtOp
-  SubtractOp
-  TanhOp
-  TorchIndexSelectOp
-  TransposeOp
-  XorOp

These are all the ops that were handled by the original linalg lowering in mlir-hlo.

Next, I plan to work towards a new input conversion pipeline that passes our existing e2e tests under `iree/tests/e2e/xla_ops`. These test input programs should be easy to convert manually to stablehlo. For simplicity of the implementation, the new input conversion will duplicate some code we already have in the mhlo input conversion. After we can pass the e2e test suite and also import larger models, we will delete the mhlo pipeline.

**Comment by kuhar** [comment]
After https://github.com/openxla/iree/pull/13439, the remaining e2e tests that do not work yet are:

- [x] 0. `exponential_fp16.mlir`
- [x] 1. `fft.mlir`
- [x] 2. `gather.mlir`
- [x] 3. `rng_normal.mlir`
- [x] 4. `scatter.mlir`

**Comment by kuhar** [comment]
- [x] We will also need a tuple flattening pass, similar to `--iree-input-type=xla`.

**Comment by jpienaar** [comment]
I see the rng lowering is tagged with this issue, https://github.com/openxla/iree/issues/13686 is related. I have a small change related - but wanted to check if you are in progress with that before going further.

**Comment by antiagainst** [comment]
FYI: @kuhar is OOO and will be back next Tuesday. I'd suggest to just go ahead (given you mentioned it's small change) if needed quickly.

**Comment by kuhar** [comment]
> I see the rng lowering is tagged with this issue, #13686 is related. I have a small change related - but wanted to check if you are in progress with that before going further.

I don't have any work in progress on this. I'd expect the solution for mhlo to be almost identical to the stablehlo one, resulting in duplicate code in the near future, before we freeze feature development in the mhlo input conversion: https://groups.google.com/g/iree-discuss/c/s6dBpDtWhtk

**Comment by kuhar** [comment]
Subtasks:

- [x] https://github.com/openxla/iree/issues/13803
- [x] https://github.com/openxla/iree/issues/13869
